math/miller_rabin.cpp

```diff
@@ -0,0 +1,177 @@
+    std::vector<T> result;
+    T temp = num;
+    while (temp > 0) {
+        result.push_back(temp % 2);
+        temp = temp / 2;
+    }
+    return result;
+}
+
+/**
+ * Function for modular exponentiation.
+ * This function is an efficient modular exponentation function.
+ * It can be used with any big integer library such as Boost multiprecision
+ * to give result any modular exponentation problem reletively quickly.
+ * @param base number being raised to a power as integer
+ * @param rev_binary_exponent reverse binary of the power the base is being
+ * raised to
+ * @param mod modulo
+ * @return r the modular exponentiation of $a^{n} \equiv r \mod{m}$ where $n$ is
+ * the base 10 representation of rev_binary_exponent and $m = mod$ parameter.
+ */
+template <typename T>
+T modular_exponentiation(T base, std::vector<T> rev_binary_exponent, T mod) {
+    if (mod == 1)
+        return 0;
+    T b = 1;
+    if (rev_binary_exponent.size() == 0)
+        return b;
+    T A = base;
+    if (rev_binary_exponent[0] == 1)
+        b = base;
+
+    for (typename std::vector<T>::iterator it = rev_binary_exponent.begin() + 1;
+         it != rev_binary_exponent.end(); ++it) {
+        A = A * A % mod;
+        if (*it == 1)
+            b = A * b % mod;
+    }
+    return b;
+}
+
+/** Function for testing the conditions that are satisfied when a number is
+ * prime.
+ * 	@param d number such that $d \cdot 2^r = n - 1$ where $n = num$
+ * parameter and r >= 1
+ * 	@param num number being tested for primality.
+ * 	@return false if n is composite and true if n is probably prime.
+ */
+template <typename T> bool miller_test(T d, T num) {
+    // random number seed
+    std::random_device rd_seed;
+    // random number generator
+    std::mt19937 gen(rd_seed());
+    // Uniformly distributed range [2, num - 2] for random numbers
+    std::uniform_int_distribution<> distribution(2, num - 2);
+    // Random number generated in the range [2, num -2].
+    T random = distribution(gen);
+    // vector for reverse binary of the power
+    std::vector<T> power = reverse_binary(d);
+    // x = random ^ d % num
+    T x = modular_exponentiation(random, power, num);
+    // miller conditions
+    if (x == 1 || x == num - 1) {
+        return true;
+    }
+
+    while (d != num - 1) {
+        x = (x * x) % num;
+        d *= 2;
+        if (x == 1) {
+            return false;
+        }
+        if (x == num - 1) {
+            return true;
+        }
+    }
+    return false;
+}
+
+/**
+ * Function that test (probabilistically) whether a given number is a prime
+ * based on the Miller-Rabin Primality Test.
+ * @param num number to be tested for primality.
+ * @param repeats number of repetitions for the test to increase probability of
+ * correct result.
+ * @return false if num is composite and true if its probably prime
+ *
+ * First we check whether the num input is less than 4, if so we can determine
```

Comment:
add `\detail` before this line. otherwise, it will be considered as part of the last known tag which is `@return`

Reply:
Added but it doesn't show outdated here.